### PR TITLE
Align executor patterns for mget and search

### PR DESCRIPTION
**Pull request by jeffmclean-cs**
This PR aligns the executor patterns used in mget and search.

It also cleans up some value vs pointer stuff.

### CODE REVIEW

File: osv2/executor.go [hunk 1]
```diff
@@ -26,45 +26,52 @@ func NewExecutor(client *opensearch.Client) *Executor {
 // will be returned.
 // An error can be returned if:
 //   - The request to OpenSearch fails
-//   - The results json cannot be unmarshalled
-func (e *Executor) MGet(ctx context.Context, req *opensearchtools.MGetRequest) (*opensearchtools.OpenSearchResponse[opensearchtools.MGetResponse], error) {
+//   - The results JSON cannot be unmarshalled
+func (e *Executor) MGet(ctx context.Context, req *opensearchtools.MGetRequest) (resp opensearchtools.OpenSearchResponse[opensearchtools.MGetResponse], err error) {
```
Comment (cs-marshall) [inline osv2/executor.go:30]: Are you just naming the return variable so you don't have to return nil? I'm not sure I follow that value of this change.

File: response.go [hunk 2]
```diff
@@ -2,11 +2,36 @@ package opensearchtools
 
 import "net/http"
 
+// RawOpenSearchResponse is a generic return type for an OpenSearch query which has meta fields common
+// to all response types as well as a generic Response field abstract across all response types.
+type RawOpenSearchResponse[T any] struct {
```
Comment (jeffmclean-cs) [inline response.go:7]: I don't really like the naming of `RawOpenSearchResponse` and `OpenSearchResponse` because "raw" is a bit misleading (makes me think of the raw response we get back from OS, but that's not really what this is).

My other thought was `OpenSearchResponse` and `OpenSearchResponseWithValidation` which is nice because it is accurate, is getting a bit long. LMKWYT.

Comment (cs-marshall) [inline response.go:7]: Why do this wrapping? I'm not sure I see the value of this extra layer. To use a `OpenSearchResponse` I now have multiple levels to get to the actual response
```
resp.RawOpenSearchResponse.Response
```
It feels a little unneeded and clunky to me

Comment (cs-marshall) [reply response.go:7]: I mentioned it in my comment. But I don't think we need this struct at all. I'm not sure I see the value this adds by separating out the validation results and having two "Response" types.

I'd maybe like it better if `OpenSearchResponse` extend `RawOpenSearchResponse` instead of having it as an attribute. But it still seems a little clunky to me.

File: osv2/search.go [hunk 3]
```diff
@@ -112,21 +112,27 @@ func (r *SearchRequest) WithQuery(q opensearchtools.Query) *SearchRequest {
...
+	return SearchRequest{
 		Query: convertedQuery,
 		Index: req.Index,
 		Size:  req.Size,
 		Sort:  req.Sort,
 	}, nil
 }
 
+// Validate validates the given SearchRequest
+func (r *SearchRequest) Validate() opensearchtools.ValidationResults {
```
Comment (jeffmclean-cs) [inline osv2/search.go:131]: Clearly this isn't doing any actual validation but 1) it's a place holder for any future validation we want to do; 2) it aligns the pattern so that we always explicitly call the .Validate() before making the call to .Do() / on to the actual open search.

File: response.go (hunk 2 above)
Comment (dylan-bourque) [reply response.go:7]: I agree with @cs-marshall.  Don't particularly like the extra type.  Why not just keep `OpenSearchResponse[T any]` as we had it? 😕

A response with no validation result will just have the `ValidationResults` field set to `nil`.  No need for a separate type, IMO.

File: osv2/executor.go (hunk 1 above)
Comment (jeffmclean-cs) [reply osv2/executor.go:30]: It has to do with trying to keep things in the stack rather than the heap which can be more optimal for performance.

File: osv2/executor.go [hunk 4]
```diff
@@ -26,45 +26,54 @@ func NewExecutor(client *opensearch.Client) *Executor {
...
-	validationRes := osv2Req.Validate()
-	if validationRes.IsFatal() {
-		return nil, opensearchtools.NewValidationError(validationRes)
+//   - The results JSON cannot be unmarshalled
+func (e *Executor) MGet(ctx context.Context, req *opensearchtools.MGetRequest) (resp opensearchtools.OpenSearchResponse[opensearchtools.MGetResponse], err error) {
+	validationResults := opensearchtools.NewValidationResults()
+
+	osv2Req, vrs := fromDomainMGetRequest(req)
+	if vrs.IsFatal() {
+		return resp, opensearchtools.NewValidationError(vrs)
```
Comment (cs-marshall) [inline osv2/executor.go:35]: do we want to add the vrs to the resp incase there are any non-fatal validation results?

File: osv2/mget.go [hunk 5]
```diff
@@ -78,45 +83,47 @@ func (m *MGetRequest) Do(ctx context.Context, client *opensearch.Client) (*MGetR
 		return nil, rErr
 	}
 
-	resp := &MGetResponse{
-		StatusCode: osResp.StatusCode,
```
Comment (cs-marshall) [inline osv2/mget.go:82]: we can also remove these fields from the `MGetResponse` struct definition as well.

File: validation.go [hunk 6]
```diff
@@ -28,21 +24,45 @@ func NewValidationResult(message string, fatal bool) ValidationResult {
...
 			return true
 		}
 	}
 
 	return false
 }
 
+// Add either adds the given [ValidationResult] to the set or does not if it already exists in the set
+func (vrs *ValidationResults) Add(vr ValidationResult) {
+	vrs.validationResults[vr] = struct{}{}
```
Comment (cs-marshall) [inline validation.go:56]: for both `Add` and `Extend` we need to nil check the `validationResults` map. Consumers aren't forced to use the `NewValidationResults()` constructor, so we should check for the edge case of a literal instantiation.

Comment (cs-marshall) [reply validation.go:56]: and then add a test for both methods trying to add/extend to a literally instantiation.

Comment (jeffmclean-cs) [reply validation.go:59]: Done, thanks!

File: osv2/mget.go (hunk 5 above)
Comment (jeffmclean-cs) [reply osv2/mget.go:82]: Nice, I missed that. Thanks

File: osv2/executor.go (hunk 4 above)
Comment (jeffmclean-cs) [reply osv2/executor.go:37]: Yes, good call. Done.

**Review by cs-marshall — APPROVED** [review]
